### PR TITLE
ARROW-17500: [Go] Kernel and KernelContext interfaces

### PR DESCRIPTION
**Pull request by zeroshade**
This implements the interface for Kernels along with a `KernelSignature` struct and the type matching for `InputType` and `OutputType` via a `TypeMatcher` interface. Along with tests for all of the above.

### CODE REVIEW

**Review by pitrou — COMMENTED** [review]
Do you plan to add any docstrings here?

**Comment by zeroshade** [comment]
@pitrou I probably should despite these all being internal and not public, so the documentation is clear for implementing kernels.

I'll go do that

**Comment by pitrou** [comment]
Is there a Go developer that can give this a review @zeroshade ?

**Comment by zeroshade** [comment]
@cyb70289 is really the only one I know who regularly comments (which is why i requested a review from him)

Unfortunately there really isn't any other regular Go developers contributing currently :disappointed:

File: go/arrow/compute/internal/kernel.go
Comment (lidavidm) [inline go/arrow/compute/internal/kernel.go:102]: Wouldn't you compare only the type IDs?

File: go/arrow/compute/internal/kernel.go
Comment (lidavidm) [inline go/arrow/compute/internal/kernel.go:88]: I guess we store Type and not the type ID since it prints out better?

File: go/arrow/compute/internal/kernel.go
Comment (lidavidm) [inline go/arrow/compute/internal/kernel.go:241]: Add a docstring for this?

File: go/arrow/compute/internal/kernel.go
Comment (lidavidm) [inline go/arrow/compute/internal/kernel.go:306]: Wouldn't you want `default` to error or panic, and break out Any as its own branch?

File: go/arrow/compute/internal/kernel.go
Comment (lidavidm) [inline go/arrow/compute/internal/kernel.go:484]: Don't you still want to validate that `types` has at least enough types to match all the non-varargs types?

File: go/arrow/compute/internal/kernel.go
Comment (zeroshade) [reply go/arrow/compute/internal/kernel.go:88]: `arrow.Type` is the  ID (analogous to `arrow::Type`. `arrow.DataType` is the whole type. So we're only storing the ID here.

Comment (lidavidm) [reply go/arrow/compute/internal/kernel.go:88]: Ah whoops :)

File: go/arrow/compute/internal/kernel.go
Comment (zeroshade) [reply go/arrow/compute/internal/kernel.go:306]: I was matching the logic in https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernel.cc#L334 

Though it might make sense to break Any as it's own branch and put a `debug.Assert` for the default

Comment (lidavidm) [reply go/arrow/compute/internal/kernel.go:306]: I guess I have the same nit there :) though it's not a big deal.

File: go/arrow/compute/internal/kernel.go
Comment (zeroshade) [reply go/arrow/compute/internal/kernel.go:484]: ah, fair. i should do a `len(types) >= (len(k.InputTypes)-1)`

File: go/arrow/compute/internal/kernel.go
Comment (zeroshade) [reply go/arrow/compute/internal/kernel.go:246]: added

File: go/arrow/compute/internal/kernel.go
Comment (zeroshade) [reply go/arrow/compute/internal/kernel.go]: added a `debug.Assert` and a separate branch for `InputAny`

**Comment by zeroshade** [comment]
I'll wait until tomorrow to merge this in order to give @cyb70289 time to comment if desired.

**Review by cyb70289 — APPROVED** [review]
LGTM

File: go/arrow/compute/internal/kernel.go
Comment (zeroshade) [reply go/arrow/compute/internal/kernel.go:93]: Sure, I'll modify it for that

**Pull request merged**
